[FLINK-25577][docs] Update GCS documentation

galenwarren opened the pull request:
## What is the purpose of the change

Update docs for GCS filesystem, including changes from [FLINK-11838](https://github.com/apache/flink/pull/15599).

## Brief change log
  - Update docs/content/docs/deployment/filesystems/gcs.md

## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no) No
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no) No
  - The serializers: (yes / no / don't know) No
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know) No
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know) No
  - The S3 file system connector: (yes / no / don't know) No

## Documentation

  - Does this pull request introduce a new feature? (yes / no) No
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented) This PR is for documentation

galenwarren commented:
This is still a work in progress; further commits will be coming soon to update the GCS docs.

I do have one question here -- at the top of the existing `gcs.md` file, it lists aliases:

    aliases:
      - /deployment/filesystems/gcs.html

The `s3.md` file has these aliases:

    aliases:
      - /deployment/filesystems/s3.html
      - /ops/filesystems/s3.html

Should `gcs.md` have a second alias as well? I'm not sure what those aliases mean, exactly.

MartijnVisser commented:
The aliases are for documentation that has been published and later has been renamed or moved. My including the original URL as alias, we prevent that users end up on a broken link

galenwarren commented:
Ah, understood. So I'll leave it as is in `gcs.md`.

galenwarren commented:
@MartijnVisser @xintongsong I think these docs are ready to go, but please let me know if you have any feedback.

Note that you can't actually follow along with the instructions to enable the `flink-gs-fs-hadoop` plugin until this [change](https://github.com/apache/flink/pull/15599#issuecomment-877470182) is merged into master; we missed this along the way and @xintongsong is going to merge it separately.

Also, if the content of the docs is good, does it need to get translated into Chinese? I'll definitely need help with that. :)

xintongsong commented on docs/content/docs/deployment/filesystems/gcs.md:
```diff
@@ -50,57 +48,40 @@ env.getCheckpointConfig().setCheckpointStorage("gs://<bucket>/<endpoint>");
 
-We have tested with `flink-shared-hadoop2-uber` version >= `2.8.5-1.8.3`.
-You can track the latest version of the [gcs-connector hadoop 2](https://storage.googleapis.com/hadoop-lib/gcs/gcs-connector-latest-hadoop2.jar).
+This file system supports the [StreamingFileSink]({{< ref "docs/connectors/datastream/streamfile_sink" >}}) and the [FileSink]({{< ref "docs/connectors/datastream/file_sink" >}}).
```
These two references are invalid, as CI complained. They are removed / renamed in FLINK-20188. I think we should now refers to "docs/content/docs/connectors/datastream/filesystem.md"

galenwarren replied:
Updated in https://github.com/apache/flink/pull/18430/commits/4d4b6ef9bf6c75d3dbbe3d64b58eb5ffcc218ebf.

xintongsong commented on docs/content/docs/deployment/filesystems/gcs.md:
```diff
@@ -50,57 +48,40 @@ env.getCheckpointConfig().setCheckpointStorage("gs://<bucket>/<endpoint>");
-You can use the following method for authentication
-* Configure via core-site.xml
-  You would need to add the following properties to `core-site.xml`
+The underlying Hadoop file system can be [configured using Hadoop's gs configuration keys](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/master/gcs/CONFIGURATION.md) by adding the configurations to your `flink-conf.yaml`.
```
Correct me if I'm wrong, I think the precise description should be:
```suggestion
The underlying Hadoop file system can be [configured using gcs-connector's Hadoop configuration keys](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/master/gcs/CONFIGURATION.md) by adding the configurations to your `flink-conf.yaml`.
```

galenwarren replied:
Yes, that's better. Good catch. Updated in https://github.com/apache/flink/pull/18430/commits/4d4b6ef9bf6c75d3dbbe3d64b58eb5ffcc218ebf.

xintongsong commented on docs/content/docs/deployment/filesystems/gcs.md:
````diff
@@ -50,57 +48,40 @@ env.getCheckpointConfig().setCheckpointStorage("gs://<bucket>/<endpoint>");
-    </property>
-  </configuration>
-  ```
+For example, Hadoop has a `fs.gs.http.connect-timeout` configuration key. If you want to change it, you need to set `gs.http.connect-timeout: xyz` in `flink-conf.yaml`. Flink will internally translate this back to `fs.gs.http.connect-timeout`. There is no need to pass configuration parameters using Hadoop's XML configuration files.
````
```suggestion
For example, gcs-connector has a `fs.gs.http.connect-timeout` configuration key. If you want to change it, you need to set `gs.http.connect-timeout: xyz` in `flink-conf.yaml`. Flink will internally translate this back to `fs.gs.http.connect-timeout`. There is no need to pass configuration parameters using Hadoop's XML configuration files.
```

galenwarren replied:
Updated in https://github.com/apache/flink/pull/18430/commits/4d4b6ef9bf6c75d3dbbe3d64b58eb5ffcc218ebf.

xintongsong commented on docs/content/docs/deployment/filesystems/gcs.md:
````diff
@@ -50,57 +48,40 @@ env.getCheckpointConfig().setCheckpointStorage("gs://<bucket>/<endpoint>");
-  ```
+| Key                                       | Description                                                                                                                                                                                                                                                                                                                                                       ...
+|-------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ ...
+| gs.writer.temporary.bucket.name           | If this property is not set, temporary blobs for in-progress writes via `RecoverableWriter` will be written to same bucket as the final file being written, prefixed with `.inprogress/`. <br><br>Set this property to choose a different bucket to hold the temporary blobs. It is recommended to choose a separate bucket in order to [assign it a TTL](https:/ ...
````
Maybe swap the first and second paragraph?

galenwarren replied:
Done in https://github.com/apache/flink/pull/18430/commits/4d4b6ef9bf6c75d3dbbe3d64b58eb5ffcc218ebf.

xintongsong reviewed (commented):
@galenwarren, thanks for working on this documentation.

Technical-wise, I think the contents look good. My major concern is regarding how we deal with the `core-site.xml` approach for authentication, as discussed [here](https://github.com/apache/flink/pull/15599#issuecomment-1018971976). Apart from that, I have only a few minor comments.

Language-wise, I don't think my English is good enough to comment. Maybe @MartijnVisser can help check this out?

galenwarren commented:
@xintongsong @MartijnVisser 

I also added back in the documentation of how to use `core-site.xml` config properties; the code for this is in the in-progress PR: https://github.com/apache/flink/pull/18489.

xintongsong reviewed (commented):
Apart from the broken link, this PR LGTM.

MartijnVisser commented on docs/content/docs/deployment/filesystems/gcs.md:
```diff
@@ -50,57 +48,66 @@ env.getCheckpointConfig().setCheckpointStorage("gs://<bucket>/<endpoint>");
-  <artifactId>flink-shaded-hadoop2-uber</artifactId>
-  <version>${flink.shared_hadoop_latest_version}</version>
-</dependency>
+This file system supports being used in [file sinks]({{< ref "docs/content/docs/connectors/datastream/filesystem.md" >}}).
```
We've recently renamed the File Sink to FileSystem. 

```suggestion
This file system can be used with the [FileSystem connector]({{< ref "docs/content/docs/connectors/datastream/filesystem.md" >}}).
```

galenwarren replied:
Fixed in https://github.com/apache/flink/pull/18430/commits/01d94ab245f6baff38f2bfcfa3ef43d52a5b2bdf.

MartijnVisser reviewed (approved):
Looks good to me, one small detail on the change we've recently made to our File Sink/FileSystem documentation.

galenwarren commented on docs/content/docs/deployment/filesystems/gcs.md:
```diff
@@ -50,57 +48,66 @@ env.getCheckpointConfig().setCheckpointStorage("gs://<bucket>/<endpoint>");
-  <artifactId>flink-shaded-hadoop2-uber</artifactId>
-  <version>${flink.shared_hadoop_latest_version}</version>
-</dependency>
+This file system supports being used in [file sinks]({{< ref "docs/content/docs/connectors/datastream/filesystem.md" >}}).
```
Fixed in https://github.com/apache/flink/pull/18430/commits/de782534bf4fb67f9d308e0d47281d2a24e18d13. Thanks for the hint on how to build locally.

galenwarren commented:
@xintongsong @MartijnVisser I've addressed your comments and I also added a warning about the use of non-supported authentication options in `core-site-xml`. I went ahead and squashed everything in case this is ready to merge, but I'm happy to make more changes if you see anything, of course. Thanks.

xintongsong reviewed (approved):
Thanks @galenwarren and @MartijnVisser. LGTM.
Once #18489 is ready, I'll merge the 2 PRs together.
